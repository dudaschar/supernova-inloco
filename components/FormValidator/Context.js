import React from 'react'

const LayoutContext = React.createContext('inloco-form-validator')

export default LayoutContext
