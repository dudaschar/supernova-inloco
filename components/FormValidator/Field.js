import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Form } from 'semantic-ui-react'

import FormValidatorContext from './Context'
import FormValidatorMessage from './Message'

export class FormValidatorField extends Component {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    name: PropTypes.string.isRequired
  }

  static contextType = FormValidatorContext

  render() {
    const { children, className, name, ...otherProps } = this.props
    const { errors = {} } = this.context
    const classes = cx('inloco-form-validator__field', className)
    return (
      <Form.Field
        className={classes}
        name={name}
        error={!!errors[name]}
        {...otherProps}>
        {children}
        <FormValidatorMessage message={errors[name]} />
      </Form.Field>
    )
  }
}

export default FormValidatorField
