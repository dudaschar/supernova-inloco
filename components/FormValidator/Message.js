import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Icon } from 'semantic-ui-react'

export class FormValidatorMessage extends Component {
  static propTypes = {
    className: PropTypes.string,
    message: PropTypes.string
  }

  render() {
    const { className, message, ...otherProps } = this.props
    const classes = cx('inloco-form-validator__message', className)
    return (
      <div className={classes} {...otherProps}>
        {message && <Icon className="error" size="big" />}
        {message}
      </div>
    )
  }
}

export default FormValidatorMessage
