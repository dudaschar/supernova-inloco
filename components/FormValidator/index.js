import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Form } from 'semantic-ui-react'

import FormValidatorContext from './Context'

export class FormValidator extends Component {
  static propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    onSubmit: PropTypes.func,
    onSubmitSuccess: PropTypes.func,
    rules: PropTypes.object.isRequired,
    value: PropTypes.object.isRequired
  }

  static Context = FormValidatorContext

  state = {
    errors: {}
  }

  render() {
    const {
      children,
      className,
      onSubmitSuccess,
      rules,
      value,
      ...otherProps
    } = this.props
    const { errors } = this.state
    const classes = cx('inloco-form-validator', className)
    return (
      <Form className={classes} {...otherProps} onSubmit={this.handleSubmit}>
        <FormValidatorContext.Provider value={{ errors }}>
          {children}
        </FormValidatorContext.Provider>
      </Form>
    )
  }

  componentDidUpdate() {
    if (this.hasErrors()) {
      this.revalidateExistingErrors()
    }
  }

  getFieldsWithError() {
    const { errors } = this.state
    return Object.keys(errors).filter(fieldName => errors[fieldName])
  }

  hasErrors() {
    return this.getFieldsWithError().length > 0
  }

  handleSubmit = (...args) => {
    const { onSubmit, onSubmitSuccess } = this.props
    onSubmit && onSubmit(...args)

    const success = this.validateAll()
    if (success && onSubmitSuccess) {
      onSubmitSuccess()
    }
  }

  revalidateExistingErrors() {
    const { errors } = this.state

    const fieldsWithError = this.getFieldsWithError()
    const { errors: newErrors } = this.runRules(fieldsWithError)

    const haveErrorsChanged = fieldsWithError.some(
      fieldName => errors[fieldName] !== newErrors[fieldName]
    )
    if (haveErrorsChanged) {
      this.setState({ errors: newErrors })
    }
  }

  runRule = fieldName => {
    const { rules, value } = this.props
    const rule = rules[fieldName]

    if (!rule.validate(value[fieldName])) {
      return typeof rule.message === 'function'
        ? rule.message({ fieldName })
        : rule.message
    }
  }

  runRules(fieldNames) {
    const errorsArray = fieldNames.map(this.runRule)
    const success = errorsArray.filter(error => !!error).length > 0

    const errors = {}
    fieldNames.forEach(
      (fieldName, index) => (errors[fieldName] = errorsArray[index])
    )

    return { errors, success }
  }

  validateAll() {
    const allFields = Object.keys(this.props.rules)
    const { errors, success } = this.runRules(allFields)
    this.setState({ errors })
    return success
  }
}

export default FormValidator
