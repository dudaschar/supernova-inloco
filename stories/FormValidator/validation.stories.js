import React, { Component } from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import { Button, Checkbox, Input, FormValidator } from '../../components'

const validate = ({ firstName, lastName, website, terms }) => {
  const errors = {}
  if (!firstName) {
    errors.firstName = 'The first name is required.'
  }
  if (!lastName) {
    errors.lastName = 'The last name is required.'
  }
  if (!website || !website.includes('.')) {
    errors.website =
      'The website is required and must have at least one dot (".").'
  }
  if (!terms) {
    errors.terms = 'The terms must be accepted.'
  }
  return errors
}

class ExampleForm extends Component {
  state = {
    value: {}
  }

  render() {
    const { value } = this.state
    return (
      <FormValidator
        validate={validate}
        value={value}
        onSubmitSuccess={action('Valid form submission!')}>
        <FormValidator.Field name="firstName">
          <label>First Name</label>
          <Input
            placeholder="First Name"
            name="firstName"
            onChange={this.handleChange}
          />
        </FormValidator.Field>
        <FormValidator.Field name="lastName">
          <label>Last Name</label>
          <Input
            placeholder="Last Name"
            name="lastName"
            onChange={this.handleChange}
          />
        </FormValidator.Field>
        <FormValidator.Field name="website">
          <label>Website</label>
          <Input
            placeholder="Website"
            label="https://"
            name="website"
            onChange={this.handleChange}
          />
        </FormValidator.Field>
        <FormValidator.Field name="terms">
          <Checkbox
            className="blue"
            label="I agree to the Terms and Conditions"
            name="terms"
            onChange={this.handleChange}
          />
        </FormValidator.Field>
        <Button type="submit">Submit</Button>
      </FormValidator>
    )
  }

  handleChange = (event, data) => {
    this.setState(({ value }) => ({
      value: {
        ...value,
        [data.name]: data.value || data.checked
      }
    }))
  }
}

storiesOf('FormValidator', module).add('basic', () => <ExampleForm />)
