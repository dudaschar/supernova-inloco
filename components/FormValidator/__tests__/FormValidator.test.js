import { shallow } from 'enzyme'
import React from 'react'
import { Form } from 'semantic-ui-react'

import FormValidator from '../'

const buildWrapper = props =>
  shallow(
    <FormValidator
      className="test-class"
      onSubmit={jest.fn()}
      onSubmitSuccess={jest.fn()}
      validate={({ age, name }) => {
        const errors = {}
        if (!name) {
          errors.name = 'Name is required'
        }
        if (!Number.isInteger(age)) {
          errors.age = 'Age should be an integer'
        }
        return errors
      }}
      value={{
        age: 29
      }}
      {...props}>
      Test Children Content
    </FormValidator>
  )

describe('FormValidator', () => {
  it('should render form and context provider with no errors at first', () => {
    expect(buildWrapper()).toMatchSnapshot()
  })

  describe('when the form is submitted', () => {
    let wrapper

    beforeEach(() => {
      wrapper = buildWrapper()
      wrapper.find(Form).simulate('submit')
    })

    it('should render a context provider with the validation errors', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call "onSubmit" prop', () => {
      const {
        props: { onSubmit }
      } = wrapper.instance()
      expect(onSubmit).toHaveBeenCalled()
    })

    describe('when the "value" changes', () => {
      it('should clear existing errors that were resolved', () => {
        wrapper.setProps({
          value: { name: 'Maira', age: 29 }
        })
        expect(wrapper).toMatchSnapshot()
      })

      it('should not show new errors', () => {
        wrapper.setProps({ value: { age: 'String age?' } })
        expect(wrapper).toMatchSnapshot()
      })
    })
  })
})
