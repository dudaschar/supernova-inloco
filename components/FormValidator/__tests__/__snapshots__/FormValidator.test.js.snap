// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormValidator should render form and context provider with no errors at first 1`] = `
<Form
  as="form"
  className="inloco-form-validator test-class"
  onSubmit={[Function]}
>
  <ContextProvider
    value={
      Object {
        "errors": Object {},
      }
    }
  >
    Test Children Content
  </ContextProvider>
</Form>
`;

exports[`FormValidator when the form is submitted should render a context provider with the validation errors 1`] = `
<Form
  as="form"
  className="inloco-form-validator test-class"
  onSubmit={[Function]}
>
  <ContextProvider
    value={
      Object {
        "errors": Object {
          "age": undefined,
          "name": "Name is required",
        },
      }
    }
  >
    Test Children Content
  </ContextProvider>
</Form>
`;

exports[`FormValidator when the form is submitted when the "value" changes should clear existing errors that were resolved 1`] = `
<Form
  as="form"
  className="inloco-form-validator test-class"
  onSubmit={[Function]}
>
  <ContextProvider
    value={
      Object {
        "errors": Object {
          "name": undefined,
        },
      }
    }
  >
    Test Children Content
  </ContextProvider>
</Form>
`;

exports[`FormValidator when the form is submitted when the "value" changes should not show new errors 1`] = `
<Form
  as="form"
  className="inloco-form-validator test-class"
  onSubmit={[Function]}
>
  <ContextProvider
    value={
      Object {
        "errors": Object {
          "age": undefined,
          "name": "Name is required",
        },
      }
    }
  >
    Test Children Content
  </ContextProvider>
</Form>
`;
